test/functional/plugin/lsp/testutil.lua

```diff
@@ -39,8 +39,7 @@ M.create_server_definition = [[
         })
         local handler = handlers[method]
         if handler then
-          local response, err = handler(method, params)
-          callback(err, response)
+          handler(method, params, callback)
```

Comment:
Is this a really necessary change (using `callback` for all the handlers) for this PR? Seems complex. This doesn't work like asynchronously, and doesn't seem quite relevant with `initialize` being deferred.

Reply:
You requested that the test simulate the more realistic async scenario. So I needed to change the test server to support async responses.

Reply:
If the tests seem too verbose, we could add a `handler_stub` testutil like:

``` lua
function M.handler_stub(response, err)
  return function (_, _, callback)
    callback(err, response)
  end
end
```

The most handler would look like this:

```
local handlers = {
  ["textDocument/codeAction"] = handler_stub({
    title = "Code Action 1",
    command = {
      title = "Command 1",
      command = "command:1",
    }
  })
}
```

Reply:
Could also support either a function with synchronous result or a table that has something like: `{type: async|sync, handler: fun}`

Let me know what you think. Otherwise I can also merge the PR as is. Changing them all to async with callback is fine by me and would also match what came up as part of https://github.com/neovim/neovim/pull/24338 (which I should get back to at some point)

Reply:
I'm happy with the current state of the PR. The callback signature didn't increase the existing test LOC so I don't think adding overloads is worth it. I've squashed the changes.